### PR TITLE
CIF/mmCIF improvements

### PR DESCRIPTION
**Pull request by Luthaf**
This PR should fix all chemfiles' side issues reported in https://github.com/MDAnalysis/mdanalysis/issues/2367#issuecomment-876065924. I'll work on the MDAnalysis side for the remaining issues.

- make the mmCIF reader more strict, and move COD/CCSD files support to CIF reader
- make `guess_format` distinguish between CIF and mmCIF files if the extension is .cif

@frodofine since you wrote the mmCIF code, could you give this a look?

### CODE REVIEW

**Comment by frodofine** [comment]
I'll try to take a look next week

File: include/chemfiles/formats/mmCIF.hpp [hunk 1]
```diff
@@ -48,9 +48,6 @@ class mmCIFFormat final: public Format {
     std::map<std::string, size_t> atom_site_map_;
     /// Map of residues, indexed by residue id and chainid.
     std::map<std::pair<std::string, int64_t>, Residue> residues_;
-    /// Set to true if the file is based on fractional coordinates.
-    /// Set to false if the file is based on cartesian coordinates.
-    bool uses_fract_;
```
Comment (frodofine) [inline include/chemfiles/formats/mmCIF.hpp:53]: Why is this removed? I think in theory mmCIF can use fractional coordinates.

File: .github/workflows/tests.yml [hunk 2]
```diff
@@ -215,7 +215,7 @@ jobs:
       - name: run tests
         run: |
           cd build
-          ctest --build-config ${{ matrix.build-type }} --parallel 2
+          ctest --output-on-failure --build-config ${{ matrix.build-type }} --parallel 2
```
Comment (frodofine) [inline .github/workflows/tests.yml:218]: Do you intend to keep "--output-on-failure"?

File: src/formats/mmCIF.cpp [hunk 3]
```diff
@@ -122,16 +122,15 @@ void mmCIFFormat::init_() {
                     trimmed.substr(1, trimmed.size() - 2).to_string() : "";
         }
 
-        if (in_loop && line_split[0].find("_atom_site") != std::string::npos) {
+        if (in_loop && line_split[0].find("_atom_site.") != std::string::npos) {
```
Comment (frodofine) [inline src/formats/mmCIF.cpp:125]: Why have you added the dot? This this just to improve specificity?

File: src/formats/mmCIF.cpp [hunk 4]
```diff
@@ -122,16 +122,15 @@ void mmCIFFormat::init_() {
...
             auto atom_label = line_split[0].substr(11).to_string();
-            to_ascii_lowercase(atom_label);
```
Comment (frodofine) [inline src/formats/mmCIF.cpp:127]: Why is this removed?

File: src/formats/mmCIF.cpp [hunk 5]
```diff
@@ -159,22 +157,14 @@ void mmCIFFormat::init_() {
...
-    // Attempt to guess how coordinates are stored
-    auto cartn_x = atom_site_map_.find("cartn_x");
-    auto fract_x = atom_site_map_.find("fract_x");
-    if (cartn_x != atom_site_map_.end()) {
-        uses_fract_ = false;
-    } else if (fract_x != atom_site_map_.end()) {
-        uses_fract_ = true;
-    } else {
-        throw format_error("CIF file does not define coordinates");
+    if (atom_site_map_.find("Cartn_x") == atom_site_map_.end()) {
```
Comment (frodofine) [inline src/formats/mmCIF.cpp:163]: Okay, so this code no longer supports fractional coordinates?

File: src/formats/mmCIF.cpp [hunk 6]
```diff
@@ -232,7 +222,7 @@ void mmCIFFormat::read(Frame& frame) {
     // and stored with optional. I don't know which solution is better.
 
     // These are required for atoms
-    auto type_symbol = atom_site_map_.find("type_symbol");
+    auto type_symbol = atom_site_map_.at("type_symbol");
```
Comment (frodofine) [inline src/formats/mmCIF.cpp:225]: Because we are guaranteed to have it!

File: src/formats/mmCIF.cpp [hunk 7]
```diff
@@ -309,20 +296,10 @@ void mmCIFFormat::read(Frame& frame) {
...
-            frame.add_atom(std::move(atom), Vector3D(x, y, z));
-        } else {
-            auto u = cif_to_double(line_split[fract_x->second].to_string());
-            auto v = cif_to_double(line_split[fract_y->second].to_string());
-            auto w = cif_to_double(line_split[fract_z->second].to_string());
-
-            auto vector = cell_.matrix() * Vector3D(u, v, w);
-            frame.add_atom(std::move(atom), vector);
-        }
+        auto x = cif_to_double(line_split[cartn_x].to_string());
```
Comment (frodofine) [inline src/formats/mmCIF.cpp:299]: Fractional coordinates no longer supported

File: src/guess_format.cpp [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+
+    if (extension.empty()) {
+        throw file_error(
+            "file at '{}' does not have an extension, provide a format name to read it",
+            path
+        );
+    }
+
+    if (extension == ".cif" && (mode == 'r' || mode == 'a')) {
+        extension = distinguish_cif_variants(path, compression).value_or(extension);
```
Comment (frodofine) [inline src/guess_format.cpp:63]: Everything makes sense now!

**Review by frodofine — APPROVED** [review]
Looks good to me, I have no changes. Most of my comments are for my understanding.

File: .github/workflows/tests.yml (hunk 2 above)
Comment (Luthaf) [reply .github/workflows/tests.yml:218]: Yes, this make debugging from CI much nicer

File: include/chemfiles/formats/mmCIF.hpp (hunk 1 above)
Comment (Luthaf) [reply include/chemfiles/formats/mmCIF.hpp:53]: No, mmCIF only defines `Cartn_x` and friends: https://mmcif.wwpdb.org/dictionaries/mmcif_pdbx_v40.dic/Categories/atom_site.html

CIF files define both `Cartn_x` and `fract_x`: https://www.iucr.org/__data/iucr/cifdic_html/1/cif_core.dic/Iatom_site_Cartn_.html / https://www.iucr.org/__data/iucr/cifdic_html/1/cif_core.dic/Iatom_site_fract_.html, but these files are handled by the GEMMI based CIF reader.

File: src/formats/mmCIF.cpp (hunk 3 above)
Comment (Luthaf) [reply src/formats/mmCIF.cpp:125]: Exactly. The issue that prompt these change was that `1AKE` contains data in the `_atom_sites.<XXX>` group, which this code would match. For mmCIF files everything related to atoms is inside the `_atom_site.<XXX>` group, hence this change.

File: src/formats/mmCIF.cpp (hunk 4 above)
Comment (Luthaf) [reply src/formats/mmCIF.cpp:127]: Just to make the code match the files. If the file uses `Cartn_x` or `pdbx_PDB_model_num`, I'd rather have the same strings in the code, without case change. Should also prevent reading data from the wrong entry if the file is malformed and contains both `cartn_x` and `Cartn_x`.

File: src/formats/mmCIF.cpp (hunk 5 above)
Comment (Luthaf) [reply src/formats/mmCIF.cpp:163]: Exactly. Support for fractional coordinates is now done with GEMMI/the CIF reader introduced in #356.

File: src/formats/mmCIF.cpp (hunk 6 above)
Comment (Luthaf) [reply src/formats/mmCIF.cpp:225]: yes

File: src/formats/mmCIF.cpp (hunk 7 above)
Comment (Luthaf) [reply src/formats/mmCIF.cpp:299]: yes

File: src/guess_format.cpp (hunk 8 above)
Comment (Luthaf) [reply src/guess_format.cpp:63]: ^-^ 🎉

**Comment by orbeckst** [comment]
Thank you!
